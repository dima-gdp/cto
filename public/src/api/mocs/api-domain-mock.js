export const mockApiDomainGetMany = {
  data: [
    {
      name: 'reg.100metrovka.com',
      authGroupId: 20,
      id: '49',
      event: {
        id: '1',
        active: true,
        defaultLanguage: 'ru',
        availableLanguages: ['ru', 'en'],
        meta: {
          languages: ['ru', 'en'],
        },
      },
    },
  ],
  meta: {
    totalCount: 1,
    pageCount: 1,
    currentPage: 1,
    perPage: 20,
  },
  status: 200,
}
