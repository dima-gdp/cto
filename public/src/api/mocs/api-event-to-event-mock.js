export const mockApiEventToEventGetMany = {
  data: [
    {
      id: '20',
      parentId: 1,
      childId: 7522,
      allowRedirect: true,
      quickRegistration: false,
      createdAt: '2020-07-09T15:29:39+0300',
      updatedAt: '2021-05-10T15:20:56+0300',
    },
    {
      id: '21',
      parentId: 1,
      childId: 3341,
      allowRedirect: true,
      quickRegistration: false,
      createdAt: '2020-07-09T15:29:39+0300',
      updatedAt: '2021-05-10T15:20:56+0300',
    },
    {
      id: '22',
      parentId: 1,
      childId: 84,
      allowRedirect: true,
      quickRegistration: true,
      createdAt: '2020-07-09T15:29:39+0300',
      updatedAt: '2021-05-10T15:20:56+0300',
    },
  ],
  meta: {
    totalCount: 3,
    pageCount: 1,
    currentPage: 1,
    perPage: 20,
  },
  status: 200,
}
