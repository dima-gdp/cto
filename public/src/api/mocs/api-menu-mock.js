export const mockHeaderMenuFixture = [
  {
    title: 'Первый уровень!',
    url: '/fewfew',
    id: 1,
    type: 'url',
    children: [
      { title: 'Второй уровень Первый уровень Первый уровень', url: '#', id: 346, type: 'url' },
      { title: 'Второй уровень', url: '#', type: 'page', id: 2564 },
    ],
  },
  { title: 'Профиль', id: 3, url: '/event/1/profile', type: 'url' },
  { title: 'Google', url: 'https://www.google.com/', id: 4, type: 'page' },
  {
    title: 'Крайняя ссылка',
    url: '#',
    id: 100,
    type: 'page',
    children: [
      { title: 'Профиль', url: '/event/1/profile', id: 253, type: 'url' },
      { title: 'Google', url: 'https://www.google.com/', id: 2114, type: 'page' },
    ],
  },
]
