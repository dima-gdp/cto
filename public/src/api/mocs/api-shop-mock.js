export const mockShopFixture = {
  id: '2',
  name: 'М(RU) - "reg.100metrovka.com Тест3_1"',
  description: 'fsdfsd илорилорлоро\nджодшлодоодл\nsdf\ndsfs\nssdfd\nsdf',
  eventId: 1,
  sellingStartedAt: '2019-12-12T00:00:00+0300',
  sellingClosedAt: '2021-12-21T18:00:00+0300',
  sellingTimeStatus: 'during',
  settings: null,
  active: true,
  paymentProviders: [5, 28],
  lang: 'ru',
}

export const mockShopSellingClosedFixture = {
  id: '2',
  name: 'М(RU) - "reg.100metrovka.com Тест3_1"',
  description: 'fsdfsd илорилорлоро\nджодшлодоодл\nsdf\ndsfs\nssdfd\nsdf',
  eventId: 1,
  sellingStartedAt: '2019-12-12T00:00:00+0300',
  sellingClosedAt: '2021-12-21T18:00:00+0300',
  sellingTimeStatus: 'after',
  settings: null,
  active: true,
  paymentProviders: [5, 28],
  lang: 'ru',
}

export const mockShopBeforeSellingFixture = {
  id: '2',
  name: 'М(RU) - "reg.100metrovka.com Тест3_1"',
  description: 'fsdfsd илорилорлоро\nджодшлодоодл\nsdf\ndsfs\nssdfd\nsdf',
  eventId: 1,
  sellingStartedAt: '2019-12-12T00:00:00+0300',
  sellingClosedAt: '2021-12-21T18:00:00+0300',
  sellingTimeStatus: 'before',
  settings: null,
  active: true,
  paymentProviders: [5, 28],
  lang: 'ru',
}

export const mockShopFixtureWithoutDescription = {
  id: '2',
  name: 'М(RU) - "reg.100metrovka.com Тест3_1"',
  eventId: 1,
  sellingStartedAt: '2019-12-12T00:00:00+0300',
  sellingClosedAt: '2021-12-21T18:00:00+0300',
  sellingTimeStatus: 'during',
  settings: null,
  active: true,
  paymentProviders: [5, 28],
  lang: 'ru',
}

export const mockProductsFixture = {
  tickets: [
    {
      id: '4',
      name: 'тестовый товар 2',
      type: 'ticket',
      description:
        'максимальная длина описания товара - 255 символов .  максимальная длина описания товара - 255 символов . максимальная длина описания товара - 255 символов . максимальная длина описания товара - 255 символов . максимальная длина описания товара - 255 симво',
      storeId: 2,
      sort: 101,
      price: 15000,
      currency: 'RUB',
      priceRub: 15000,
      isCurrencyAllowed: false,
      isRateBlocked: false,
      settings: [],
      active: true,
      limitCount: null,
      sellingCount: null,
      orderLimitCount: null,
      sellingStartedAt: null,
      sellingClosedAt: null,
      sellingTimeStatus: 'during',
    },
  ],
  others: [
    {
      id: '7',
      name: 'тестовый товар',
      type: 'other',
      description: 'ура товарищи! айтишник может создавать товары',
      storeId: 2,
      sort: 101,
      price: 182,
      currency: 'RUB',
      priceRub: 182,
      isCurrencyAllowed: false,
      isRateBlocked: false,
      settings: [],
      active: true,
      limitCount: null,
      sellingCount: null,
      orderLimitCount: null,
      sellingStartedAt: null,
      sellingClosedAt: '2021-12-20T00:00:00+0300',
      sellingTimeStatus: 'during',
    },
  ],
}

export const mockCurrenciesFixture = [
  {
    label: 'USD',
    value: 74.37,
    id: 'USD',
  },
  {
    label: 'EUR',
    value: 88.14,
    id: 'EUR',
  },
]

export const mockCreateOrderFixture = {
  storeId: '2',
  userId: 19579,
  status: 'created',
  storeType: 'event',
  createdAt: 1633681792,
  updatedAt: 1633681792,
  id: '4743',
  currency: 'RUB',
  cost: '15000',
}
