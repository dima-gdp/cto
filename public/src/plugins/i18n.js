import Vue from 'vue'
import VueI18n from 'vue-i18n'
import { kebabToCamel, snakeToCamel } from '@/utils'
import iviewEN from '@/assets/vendor/view-design/src/locale/lang/en-US'
import iviewRU from '@/assets/vendor/view-design/src/locale/lang/ru-RU'
import LangService from '@/domain/services/lang-service'
import errorPageLocales from '@/locales/error'

Vue.use(VueI18n)

const localMessages = {
  ru: {
    ...iviewRU,
    error: errorPageLocales.ru,
  },
  en: {
    ...iviewEN,
    error: errorPageLocales.en,
  },
}

const i18n = new VueI18n({
  locale: 'ru',
  fallbackLocale: 'ru',
  messages: localMessages,
})

export const EMPTY_LOCALE_MESSAGE = ' '

export function translateFn(key, ...args) {
  // в таблице i18n все должно быть в camelCase
  const prepKey = kebabToCamel(snakeToCamel(key))

  return i18n.te(`${prepKey}`) ? i18n.t(`${prepKey}`, ...args) : EMPTY_LOCALE_MESSAGE
}

Vue.mixin({
  methods: {
    $tr(key, ...args) {
      return translateFn(key, ...args)
    },

    /**
     * @param route
     * @returns {Promise<void>}
     */
    async $toLocalePath(route) {
      await LangService.toLocalePath(route)
    },

    /**
     * @param route
     * @returns route
     */
    localePath(route) {
      return { ...route, query: { ...route.query, lang: this.$i18n.locale } }
    },
  },
})

export default i18n
