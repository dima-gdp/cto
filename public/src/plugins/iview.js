import Vue from 'vue'
import iview from '@/assets/vendor/view-design/src'

Vue.use(iview)
