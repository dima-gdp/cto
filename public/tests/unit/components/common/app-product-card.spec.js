import AppProductCard from '@/components/common/app-product-card'
import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import ShopService from '@/domain/services/shop-service'
import iview from '@/assets/vendor/view-design/src'
import Button from '@/assets/vendor/view-design/src/components/button/button.vue'
import { mockProductsFixture } from '@/api/mocs/api-shop-mock'
import AppTextEllipsis from '@/components/common/app-text-ellipsis'
import AppShopPriceView from '@/components/common/shop/app-shop-price-view'
import AppShopAppBasketIcon from '@/components/common/shop/app-shop-basket-icon'
import InputNumber from '@/assets/vendor/view-design/src/components/input-number/input-number.vue'
import { EMPTY_LOCALE_MESSAGE } from '@/plugins/i18n'

Vue.use(iview)
Vue.use(Vuex)

const I18N_KEYS = {
  'shop.btn.delete': 'shop.btn.delete',
  'shop.btn.order': 'shop.btn.order',
  'shop.product.inBasket': 'shop.product.inBasket',
  'shop.product.available': 'shop.product.available',
  'shop.product.ticket.alreadySelected': 'shop.product.ticket.alreadySelected',
  'shop.product.other.outOfStock': 'shop.product.other.outOfStock',
}

const mockRouterPush = jest.fn()
const mockSuccessMessage = jest.fn()

const mock$tr = jest.fn().mockImplementation((key) => {
  return I18N_KEYS[key] || EMPTY_LOCALE_MESSAGE
})

describe('@/components/common/shop/app-shop-basket.vue', () => {
  const TICKET = mockProductsFixture.tickets[0]
  const OTHER = mockProductsFixture.others[0]

  let wrapper
  let store

  const createComponent = ({ product, productsInBasket, isSomeTicketSelected }) => {
    store = new Vuex.Store({
      modules: {
        shop: {
          namespaced: true,
          state: {
            productsInBasket,
          },
          getters: {
            isSomeTicketSelected: () => isSomeTicketSelected,
          },
        },
      },
    })

    wrapper = shallowMount(AppProductCard, {
      store,
      propsData: {
        product,
      },
      mocks: {
        $route: {
          path: '/event/1/store',
        },
        $router: {
          push: mockRouterPush,
        },
        $Message: {
          success: mockSuccessMessage,
        },
      },
      mixins: [
        {
          methods: {
            $tr: mock$tr,
          },
        },
      ],
    })
  }

  const getProductCardBottomHTML = () => wrapper.find('.shop-ticket__bottom-info').html()

  afterEach(() => {
    wrapper.destroy()
  })

  describe('AppTextEllipsis', () => {
    it('Отрисовывает заголовок товара', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })
      const TitleComponent = wrapper
        .findAllComponents(AppTextEllipsis)
        .wrappers.filter((el) => el.props('text') === TICKET.name)[0]

      expect(TitleComponent.isVisible()).toBeTruthy()
    })

    it('Передает компоненту заголовка товара(AppTextEllipsis) lines: 2', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })
      const TitleComponent = wrapper
        .findAllComponents(AppTextEllipsis)
        .wrappers.filter((el) => el.props('text') === TICKET.name)[0]

      expect(TitleComponent.props('lines')).toBe(2)
    })

    it('Отрисовывает описание товара', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })
      const DescriptionComponent = wrapper
        .findAllComponents(AppTextEllipsis)
        .wrappers.filter((el) => el.props('text') === TICKET.description)[0]

      expect(DescriptionComponent.isVisible()).toBeTruthy()
    })

    it('Передает компоненту описания товара(AppTextEllipsis) lines: 2', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })
      const DescriptionComponent = wrapper
        .findAllComponents(AppTextEllipsis)
        .wrappers.filter((el) => el.props('text') === TICKET.description)[0]

      expect(DescriptionComponent.props('lines')).toBe(2)
    })
  })

  describe('AppShopPriceView', () => {
    it('Отрисовывает цену(AppShopPriceView)', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(AppShopPriceView).isVisible()).toBeTruthy()
    })

    it('Передает в AppShopPriceView currency продукта)', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      const PriceComponent = wrapper.findComponent(AppShopPriceView)
      expect(PriceComponent.props('currency')).toBe(TICKET.currency)
    })

    it('Передает в AppShopPriceView price продукта)', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      const PriceComponent = wrapper.findComponent(AppShopPriceView)
      expect(PriceComponent.props('price')).toBe(TICKET.price)
    })

    it('Передает в AppShopPriceView price-rub продукта)', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      const PriceComponent = wrapper.findComponent(AppShopPriceView)
      expect(PriceComponent.props('priceRub')).toBe(TICKET.priceRub)
    })
  })

  describe('Если товар является ticket', () => {
    it('Если товар является ticket, то отрисовываем кнопку купить/удалить', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(Button).isVisible()).toBeTruthy()
    })

    it('Если товар является ticket, и текущий товар выбран то в кнопку удалить передаем соответствующий текст из i18n', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(Button).vm.$slots.default[0].text.trim()).toBe(
        I18N_KEYS['shop.btn.delete'],
      )
    })

    it('Если товар является ticket, и текущий товар не выбран то в кнопку добавить передаем соответствующий текст из i18n', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(Button).vm.$slots.default[0].text.trim()).toBe(
        I18N_KEYS['shop.btn.order'],
      )
    })

    it('Передаем в кнопку купить/удалить long: true', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

      expect(wrapper.findComponent(Button).props('long')).toBe(true)
    })

    it('Если добавили тикет в корзину, и это не текущий тикет, то блокируем кнопку купить/удалить', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

      expect(wrapper.findComponent(Button).props('disabled')).toBe(true)
    })

    it('Если добавили тикет в корзину, и это не текущий тикет, то передаем type: primary в кнопку купить/удалить', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

      expect(wrapper.findComponent(Button).props('type')).toBe('primary')
    })

    it('Если добавили тикет в корзину, и это текущий тикет, то передаем type: error в кнопку купить/удалить', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

      expect(wrapper.findComponent(Button).props('type')).toBe('error')
    })

    it('Если добавили тикет в корзину, и это текущий тикет, то по клику на кнопку удалить, емитим событие change-amount с нулем', async () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })
      await wrapper.findComponent(Button).vm.$emit('click')

      expect(wrapper.emitted('change-amount')[0][0]).toBe(0)
    })

    it('Если добавили тикет в корзину, и это не текущий тикет, то по клику на кнопку удалить, емитим событие change-amount c единицей', async () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })
      await wrapper.findComponent(Button).vm.$emit('click')

      expect(wrapper.emitted('change-amount')[0][0]).toBe(1)
    })

    describe('shop-ticket__bottom-info', () => {
      it('Если товар является ticket, и не один тикет не в корзине, и количество товара не бесконечное', () => {
        const AVAILABLE_REMAINS = 5
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(AVAILABLE_REMAINS)
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })

      it('Если товар является ticket, и текущий товар в корзине, и количество товара не бесконечное', () => {
        const AVAILABLE_REMAINS = 5
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(AVAILABLE_REMAINS)
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })

      it('Если товар является ticket, и текущий товар в корзине, и количество товара бесконечное', () => {
        const INFINITY_REMAINS = Infinity
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(INFINITY_REMAINS)
        createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })

      it('Если товар является ticket, и текущий товар не в корзине, и какой-то тикет в корзине', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: true })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })
    })
  })

  describe('Если товар является other', () => {
    describe('Если это текущий other', () => {
      it('Если это текущий other, то отрисовываем InputNumber', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).isVisible()).toBeTruthy()
      })

      it('Если это текущий other, то не отрисовываем кнопку купить/удалить', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).exists()).toBeFalsy()
      })

      it('Если это текущий other, то передаем в InputNumber value равное количеству товара', () => {
        const AMOUNT = 3
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        jest.spyOn(ShopService, 'getAmountItemById').mockReturnValue(AMOUNT)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).props('value')).toBe(AMOUNT)
      })

      it('Если это текущий other, то передаем в InputNumber editable: false', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).props('editable')).toBe(false)
      })

      it('Если это текущий other, то передаем в InputNumber controls-outside: true', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).props('controlsOutside')).toBe(true)
      })

      it('Если это текущий other, то передаем в InputNumber min: 0', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).props('min')).toBe(0)
      })

      it('Если это текущий other, то передаем в InputNumber max равный максимально доступному для продажи количеству', () => {
        const MAX_AMOUNT = 10
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(MAX_AMOUNT)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).props('max')).toBe(MAX_AMOUNT)
      })

      it('Если это текущий other, то при изменении value InputNumber, емитим наверх событие "change-amount" с нагрузкой', async () => {
        const PAYLOAD_FOR_CHANGE_AMOUNT = 'payload'
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        jest.spyOn(ShopService, 'getAmountItemById').mockReturnValue(1)

        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })
        await wrapper.findComponent(InputNumber).vm.$emit('input', PAYLOAD_FOR_CHANGE_AMOUNT)

        expect(wrapper.emitted('change-amount')[0][0]).toBe(PAYLOAD_FOR_CHANGE_AMOUNT)
      })
    })

    describe('Если это не текущий other', () => {
      it('Если это не текущий other, то не отрисовываем InputNumber', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(InputNumber).exists()).toBeFalsy()
      })

      it('Если это не текущий other, то отрисовываем кнопку купить/удалить', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).isVisible()).toBeTruthy()
      })

      it('Если это не текущий other, то отрисовываем кнопку купить и передаем соответствующий текст из i18n', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).vm.$slots.default[0].text.trim()).toBe(
          I18N_KEYS['shop.btn.order'],
        )
      })

      it('Передаем в кнопку купить long: true', () => {
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).props('long')).toBe(true)
      })

      it('Если пользователь может купить товаров меньше 1, то блокируем кнопку купить', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(0)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).props('disabled')).toBe(true)
      })

      it('Если пользователь может купить товаров больше 0, то не блокируем кнопку купить', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(5)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).props('disabled')).toBe(false)
      })

      it('Передаем в кнопку купить type: primary', () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(wrapper.findComponent(Button).props('type')).toBe('primary')
      })

      it('По клику на кнопку купить, емитим событие change-amount c единицей', async () => {
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })
        await wrapper.findComponent(Button).vm.$emit('click')

        expect(wrapper.emitted('change-amount')[0][0]).toBe(1)
      })
    })

    describe('shop-ticket__bottom-info', () => {
      it('Если товар является other, и юзер может купить больше нуля товара, и если известно максимальное количество товара', () => {
        const AVAILABLE_REMAINS = 5
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(AVAILABLE_REMAINS)
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })

      it('Если товар является other, и юзер не может купить товар', () => {
        const ZERO_REMAINS = 0
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(ZERO_REMAINS)
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })

      it('Если товар является other, и юзер может купить товара больше 0, и количество товара бесконечно', () => {
        const INFINITY_REMAINS = Infinity
        jest.spyOn(ShopService, 'getRemainsByProduct').mockReturnValue(INFINITY_REMAINS)
        jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
        createComponent({ product: OTHER, productsInBasket: [], isSomeTicketSelected: true })

        expect(getProductCardBottomHTML()).toMatchSnapshot()
      })
    })
  })

  describe('AppShopAppBasketIcon', () => {
    it('Отрисовываем AppShopAppBasketIcon', () => {
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(AppShopAppBasketIcon).isVisible()).toBeTruthy()
    })

    it('Передаем в AppShopAppBasketIcon is-highlighted: true, если продукт добавлен в корзину', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(true)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(AppShopAppBasketIcon).props('isHighlighted')).toBe(true)
    })

    it('Передаем в AppShopAppBasketIcon is-highlighted: true, если продукт добавлен в корзину', () => {
      jest.spyOn(ShopService, 'isItemAlreadySelected').mockReturnValue(false)
      createComponent({ product: TICKET, productsInBasket: [], isSomeTicketSelected: false })

      expect(wrapper.findComponent(AppShopAppBasketIcon).props('isHighlighted')).toBe(false)
    })
  })
})
