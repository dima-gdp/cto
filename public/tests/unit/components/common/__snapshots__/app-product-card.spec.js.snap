// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/common/shop/app-shop-basket.vue Если товар является other shop-ticket__bottom-info Если товар является other, и юзер может купить больше нуля товара, и если известно максимальное количество товара 1`] = `
<div class="shop-ticket__bottom-info">
  <div class="shop-ticket__basket-info">
    shop.product.inBasket: 1
  </div>
  <div class="shop-ticket__access-info">
    shop.product.available: 4
  </div>
</div>
`;

exports[`@/components/common/shop/app-shop-basket.vue Если товар является other shop-ticket__bottom-info Если товар является other, и юзер может купить товара больше 0, и количество товара бесконечно 1`] = `
<div class="shop-ticket__bottom-info">
  <div class="shop-ticket__basket-info">
    shop.product.inBasket: 1
  </div>
  <!---->
</div>
`;

exports[`@/components/common/shop/app-shop-basket.vue Если товар является other shop-ticket__bottom-info Если товар является other, и юзер не может купить товар 1`] = `
<div class="shop-ticket__bottom-info">
  <div>shop.product.other.outOfStock</div>
</div>
`;

exports[`@/components/common/shop/app-shop-basket.vue Если товар является ticket shop-ticket__bottom-info Если товар является ticket, и не один тикет не в корзине, и количество товара не бесконечное 1`] = `
<div class="shop-ticket__bottom-info">
  <div class="shop-ticket__basket-info">
    shop.product.inBasket: 0
  </div>
  <div class="shop-ticket__access-info">
    shop.product.available: 5
  </div>
</div>
`;

exports[`@/components/common/shop/app-shop-basket.vue Если товар является ticket shop-ticket__bottom-info Если товар является ticket, и текущий товар в корзине, и количество товара бесконечное 1`] = `
<div class="shop-ticket__bottom-info">
  <div class="shop-ticket__basket-info">
    shop.product.inBasket: 0
  </div>
  <!---->
</div>
`;

exports[`@/components/common/shop/app-shop-basket.vue Если товар является ticket shop-ticket__bottom-info Если товар является ticket, и текущий товар в корзине, и количество товара не бесконечное 1`] = `
<div class="shop-ticket__bottom-info">
  <div class="shop-ticket__basket-info">
    shop.product.inBasket: 0
  </div>
  <div class="shop-ticket__access-info">
    shop.product.available: 5
  </div>
</div>
`;

exports[`@/components/common/shop/app-shop-basket.vue Если товар является ticket shop-ticket__bottom-info Если товар является ticket, и текущий товар не в корзине, и какой-то тикет в корзине 1`] = `
<div class="shop-ticket__bottom-info">
  <div>shop.product.ticket.alreadySelected</div>
</div>
`;
